sdk/core/Azure.Core.TestFramework/src/TestEnvironment.cs

```diff
@@ -294,6 +308,11 @@ protected string GetRecordedVariable(string name)
         protected string GetRecordedVariable(string name, Action<RecordedVariableOptions> options)
         {
             var value = GetRecordedOptionalVariable(name, options);
+            if (value == null)
+            {
+                BootStrapTestResources();
```

Comment:
Why do we need these in Get* methods?

Reply:
How else would we know if bootstrapping is required?

Reply:
Check that RG exists.

Reply:
That doesn't work - you could have an existent RG and still not have the resources.

Reply:
You can have an existing variable and no resource as well.

Reply:
I think we shouldn't target the test-resource.json editing scenario with this feature.

Reply:
Move the check to EnsureValue?

Reply:
No, because you need to recheck after running the script.

Reply:
Recheck what?

Reply:
The value for the variable.

Reply:
Can we re-check it in EnsureValue?

Reply:
I did this initially, but it we need to recheck it with two different methods, so we'd need to pass a flag or a func param.
Also, it would be calling itself recursively.

Reply:
Could this logic be short circuited and occur during construction when `ParseEnvironmentFile` fails to find an env file? While it's possible that the environment can exist without the file in place, in practice it would usually be there on Windows and avoids the check in the Get methods.

Reply:
In that case, what would happen if I prefer to set environment variables on Windows and don't have a file?

Reply:
![](https://media.giphy.com/media/j2pOFyuTJqWj9S5qdE/giphy.gif?cid=ecf05e47k2vchr3in8vwgjab0jq22w9b53n6ox19jb85swzn&rid=giphy.gif&ct=g)

Reply:
Yeah, that is the problem - we still support env vars.

Reply:
If someone is using env vars, we would end up just creating an extra set of resources for them. I suppose it would be a one-time occurrence (per service), but still not the ideal behavior.

Reply:
> If someone is using env vars, we would end up just creating an extra set of resources for them. I suppose it would be a one-time occurrence (per service), but still not the ideal behavior.

Assuming the resources already created were not done with custom arguments to the script, wouldn't running the New-TestResources script again essentially do nothing?

Reply:
If the resource group already exists, I think you could choose to update existing RG or not. If you choose not to update, the env file will not get created and you will always hit this step. If you do update, the script may or may not be successful depending on the naming scheme used for the individual resources, i.e. if it is just using the basename, then the resources would already exist and the script would fail.

Reply:
>  i.e. if it is just using the basename, then the resources would already exist and the script would fail.

Maybe I'm misunderstanding the scenario, but I thought update would succeed if update is selected and the resources already exist.

Reply:
Nope, it will still try to create the resources which would fail.

Reply:
Hmm - I just tried running it twice and it worked fine after this:
```
Confirm
Provided resource group already exists. Are you sure you want to update it?
[Y] Yes  [N] No  [S] Suspend  [?] Help (default is "Y"): Y
```

Reply:
I just tested it again and am seeing the same behavior as @christothes - the resources are not recreated if they already exist and the env file is recreated.

In terms of changing the behavior, to only check at startup, there a few tradeoffs to consider:
- users using env vars would be prompted unnecessarily. They could still just create the env file to avoid being prompted again in the future. Or as @christothes suggested, if this is enough of a problem, we could introduce another env var that, when set, will skip the bootstrap step.
- if a user is using env vars, and they do create the env file, it is possible that it could break their scenario due to how the variable lookup works - prefixed variables take precedence over non-prefixed and the env file would add prefixed variables. It wouldn't really break the scenario as much as cause the env var resources to no longer be used. 

I'm fine with either approach (the current approach, or changing to check once per test environment initialization)
What do others think?

Reply:
I would love it if we only checked on startup but I'm afraid that it would make the lives of people that use env vars even more miserable. I wonder if we can check for some well know variables like `AZURE_CLIENT_ID` and disable this behavior. Maybe we should ping our channel/partner DL and ask for people that use env vars.

Reply:
The concern that I have is the experience for external contributors.   If I'm trying to get myself running and I encounter this because I'm on Linux or am using the documented/supported approach of environment variables on Windows,  the experience isn't awesome.    Will I bother to let you know or will I just get frustrated and give up without a word?

I'd prefer that we optimize for not prompting unless we can prove that the environment isn't suitable to run, rather than optimizing for the common use case of the SDK team.

Reply:
The prompt will only happen on Windows.

Reply:
Fair enough, but I think the overall concern stands.

Reply:
Yep, if you are using env vars you may find the behavior annoying if we switch to check for the existence of an env file.

Reply:
If the env vars aren't detected (or the file not found - normal behavior now) why not just point them at the instructions on how to run New-TestResources? Adding extra time to check all the time for resources existing will slow tests down even further and some of them cost by the hour. You also can't discern much from incremental deployments. Some resources might support it, while others don't because of the configuration they are in - like resources configured for Customer-Managed Keys (CMK) like sdk/search. You can't redeploy resources in that state incrementally - it messes everything up.